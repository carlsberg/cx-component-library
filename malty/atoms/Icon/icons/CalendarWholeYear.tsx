import * as React from 'react';

function CalendarWholeYear() {
  return (
    <path
      d="M17 1a1 1 0 01.993.883L18 2v1h2a3 3 0 012.995 2.823L23 6v12.167c0 1.988-1.172 3.714-2.827 3.827L20 22H4c-1.688 0-2.916-1.66-2.996-3.629L1 18.167V6a3 3 0 012.824-2.995L4 3h2V2a1 1 0 011.993-.117L8 2v1h8V2a1 1 0 011-1zM6 5H4a1 1 0 00-.993.883L3 6v12.167c0 1.008.476 1.742.918 1.825L4 20h16c.435 0 .937-.675.995-1.648l.005-.185V6c0-.513-.386-.936-.883-.994L20 5h-2v1a1 1 0 01-1.993.117L16 6.001V5H8v1a1 1 0 01-1.993.117L6 6.001zm3.81 2.407l.094.083 2.803 2.802a1 1 0 01.084 1.32l-.083.094-2.803 2.804a1 1 0 01-1.498-1.32l.083-.094L9.585 12H8.491a1.974 1.974 0 00-.147 3.943l.147.006H15.3a1.974 1.974 0 00.147-3.943L15.3 12a1 1 0 010-2 3.974 3.974 0 01.198 7.943l-.198.005H8.49a3.974 3.974 0 01-.198-7.944L8.491 10l1.094-.001L8.49 8.905a1 1 0 01-.083-1.32l.083-.095a1 1 0 011.32-.083z"
      fillRule="evenodd"
    />
  );
}

export default CalendarWholeYear;
