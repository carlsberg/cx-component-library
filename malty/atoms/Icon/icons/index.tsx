export { default as AddContent } from './AddContent';
export { default as AddedDocument } from './AddedDocument';
// export { default as AddEvent } from './AddEvent';
// export { default as AddProject } from './AddProject';
// export { default as Alert } from './Alert';
// export { default as AlertConnection } from './AlertConnection';
// export { default as AppDropbox } from './AppDropbox';
// export { default as AppFacebook } from './AppFacebook';
// export { default as AppGithub } from './AppGithub';
// export { default as AppInstagram } from './AppInstagram';
// export { default as AppLinkedin } from './AppLinkedin';
// export { default as AppPowerbi } from './AppPowerbi';
// export { default as AppSkype } from './AppSkype';
// export { default as Arrange } from './Arrange';
// export { default as ArrowSmallDown } from './ArrowSmallDown';
// export { default as ArrowSmallLeft } from './ArrowSmallLeft';
// export { default as ArrowSmallRight } from './ArrowSmallRight';
// export { default as ArrowSmallUp } from './ArrowSmallUp';
// export { default as At } from './At';
// export { default as Attachment } from './Attachment';
// export { default as Badge } from './Badge';
// export { default as Balance } from './Balance';
// export { default as Bell } from './Bell';
// export { default as BellNotification } from './BellNotification';
// export { default as Bold } from './Bold';
// export { default as Book } from './Book';
// export { default as Bookmark } from './Bookmark';
// export { default as BookmarkFilled } from './BookmarkFilled';
// export { default as Bottle } from './Bottle';
// export { default as BottleCraft } from './BottleCraft';
// export { default as Calendar } from './Calendar';
// export { default as CalendarAutumn } from './CalendarAutumn';
// export { default as CalendarConfirm } from './CalendarConfirm';
// export { default as CalendarEvents } from './CalendarEvents';
// export { default as CalendarSpring } from './CalendarSpring';
// export { default as CalendarSummer } from './CalendarSummer';
// export { default as CalendarWholeYear } from './CalendarWholeYear';
// export { default as CalendarWinter } from './CalendarWinter';
// export { default as Camera } from './Camera';
// export { default as Can } from './Can';
// export { default as Card } from './Card';
// export { default as CardPlus } from './CardPlus';
// export { default as CarlsbergFilled } from './CarlsbergFilled';
// export { default as CarlsbergOutlined } from './CarlsbergOutlined';
// export { default as CarretDown } from './CarretDown';
// export { default as CarretLeft } from './CarretLeft';
// export { default as CarretMove } from './CarretMove';
// export { default as CarretRight } from './CarretRight';
// export { default as CarretUp } from './CarretUp';
// export { default as Cart } from './Cart';
// export { default as CartAdded } from './CartAdded';
// export { default as CartLabel } from './CartLabel';
// export { default as CartNotification } from './CartNotification';
// export { default as Change } from './Change';
// export { default as Check } from './Check';
// export { default as CheckboxCheck } from './CheckboxCheck';
// export { default as CheckboxCheckAlternate } from './CheckboxCheckAlternate';
// export { default as CheckboxCheckSquare } from './CheckboxCheckSquare';
// export { default as CheckboxEmpty } from './CheckboxEmpty';
// export { default as ChevronDown } from './ChevronDown';
// export { default as ChevronLeft } from './ChevronLeft';
// export { default as ChevronRight } from './ChevronRight';
// export { default as ChevronUp } from './ChevronUp';
// export { default as Clock } from './Clock';
// export { default as ClockFilled } from './ClockFilled';
// export { default as Clone } from './Clone';
// export { default as Close } from './Close';
// export { default as CoinDollar } from './CoinDollar';
// export { default as CoinEuro } from './CoinEuro';
// export { default as Compass } from './Compass';
// export { default as CompassFilled } from './CompassFilled';
// export { default as Connected } from './Connected';
// export { default as Contract } from './Contract';
// export { default as Cookies } from './Cookies';
// export { default as Coupon } from './Coupon';
// export { default as Customer } from './Customer';
// export { default as CustomerAdd } from './CustomerAdd';
// export { default as CustomerQuestion } from './CustomerQuestion';
// export { default as Customers } from './Customers';
// export { default as DataTransfer } from './DataTransfer';
// export { default as DiagonalDownLeft } from './DiagonalDownLeft';
// export { default as DiagonalDownRight } from './DiagonalDownRight';
// export { default as DiagonalUpLeft } from './DiagonalUpLeft';
// export { default as DiagonalUpRight } from './DiagonalUpRight';
// export { default as Diploma } from './Diploma';
// export { default as Diploma2 } from './Diploma2';
// export { default as Download } from './Download';
// export { default as Draughtmaster } from './Draughtmaster';
// export { default as Edit } from './Edit';
// export { default as Email } from './Email';
// export { default as Expand } from './Expand';
// export { default as Eye } from './Eye';
// export { default as EyeFill } from './EyeFill';
// export { default as File } from './File';
// export { default as FilesCsv } from './FilesCsv';
// export { default as FilesJson } from './FilesJson';
// export { default as FilesPdf } from './FilesPdf';
// export { default as FilesXls } from './FilesXls';
// export { default as Filters } from './Filters';
// export { default as Flash } from './Flash';
// export { default as GlassGoblet } from './GlassGoblet';
// export { default as GlassMug } from './GlassMug';
// export { default as GlassWeissbier } from './GlassWeissbier';
// export { default as Globe } from './Globe';
// export { default as Headline } from './Headline';
// export { default as Heart } from './Heart';
// export { default as HeartFilled } from './HeartFilled';
// export { default as Help } from './Help';
// export { default as HideMenu } from './HideMenu';
// export { default as History } from './History';
// export { default as Home } from './Home';
// export { default as Hop } from './Hop';
// export { default as Image } from './Image';
// export { default as Information } from './Information';
// export { default as Italic } from './Italic';
// export { default as ItemAdd } from './ItemAdd';
// export { default as ItemCheck } from './ItemCheck';
// export { default as ItemCheckDefault } from './ItemCheckDefault';
// export { default as ItemCheckFilled } from './ItemCheckFilled';
// export { default as ItemClose } from './ItemClose';
// export { default as ItemCloseFilled } from './ItemCloseFilled';
// export { default as Library } from './Library';
// export { default as Loading } from './Loading';
// export { default as LocationIp } from './LocationIp';
// export { default as Lock } from './Lock';
// export { default as LockOpen } from './LockOpen';
// export { default as Map } from './Map';
// export { default as Market } from './Market';
// export { default as MediaPause } from './MediaPause';
// export { default as MediaPlay } from './MediaPlay';
// export { default as MediaSound } from './MediaSound';
// export { default as MediaSoundOff } from './MediaSoundOff';
// export { default as MediaStop } from './MediaStop';
// export { default as Menu } from './Menu';
// export { default as MenuLeft } from './MenuLeft';
// export { default as MenuRight } from './MenuRight';
// export { default as Minus } from './Minus';
// export { default as Mobile } from './Mobile';
// export { default as More } from './More';
// export { default as Mouse } from './Mouse';
// export { default as Notepad } from './Notepad';
// export { default as Outlet } from './Outlet';
// export { default as OutletPlus } from './OutletPlus';
// export { default as Pack } from './Pack';
// export { default as PackageKeg } from './PackageKeg';
// export { default as PackageWholesaler } from './PackageWholesaler';
// export { default as Paint } from './Paint';
// export { default as Pallet } from './Pallet';
// export { default as Pencil } from './Pencil';
// export { default as PhoneClassic } from './PhoneClassic';
// export { default as Pilsner } from './Pilsner';
// export { default as Pin } from './Pin';
// export { default as PinActive } from './PinActive';
// export { default as PinChurn } from './PinChurn';
// export { default as PinClosed } from './PinClosed';
// export { default as PinInactive } from './PinInactive';
// export { default as PinIndirect } from './PinIndirect';
// export { default as PinLabeling } from './PinLabeling';
// export { default as PinParked } from './PinParked';
// export { default as PinProspect } from './PinProspect';
// export { default as PinSeveral } from './PinSeveral';
// export { default as Pint } from './Pint';
// export { default as Plus } from './Plus';
// export { default as Power } from './Power';
// export { default as RadiobuttonOff } from './RadiobuttonOff';
// export { default as RadiobuttonOn } from './RadiobuttonOn';
// export { default as Refresh } from './Refresh';
// export { default as Restaurant } from './Restaurant';
// export { default as Retry } from './Retry';
// export { default as Salesrep } from './Salesrep';
// export { default as Save } from './Save';
// export { default as Search } from './Search';
// export { default as SearchUser } from './SearchUser';
// export { default as Sent } from './Sent';
// export { default as Settings } from './Settings';
// export { default as Share } from './Share';
// export { default as ShieldExclamation } from './ShieldExclamation';
// export { default as ShieldSecurity } from './ShieldSecurity';
// export { default as Shop } from './Shop';
// export { default as SignOut } from './SignOut';
// export { default as SpeachBalloon } from './SpeachBalloon';
// export { default as SpeachBalloonComplaint } from './SpeachBalloonComplaint';
// export { default as SpeachBalloonLivechat } from './SpeachBalloonLivechat';
// export { default as Star } from './Star';
// export { default as StarFilled } from './StarFilled';
// export { default as Survey2 } from './Survey2';
// export { default as Sync } from './Sync';
// export { default as Tap } from './Tap';
// export { default as TapGlass } from './TapGlass';
// export { default as TechnicalSupport } from './TechnicalSupport';
// export { default as Text } from './Text';
// export { default as ThirdParty } from './ThirdParty';
// export { default as ThumbDown } from './ThumbDown';
// export { default as ThumbUp } from './ThumbUp';
// export { default as ToggleOff } from './ToggleOff';
// export { default as ToggleOn } from './ToggleOn';
// export { default as Trash } from './Trash';
// export { default as Trophy } from './Trophy';
// export { default as Truck } from './Truck';
// export { default as Tulipe } from './Tulipe';
// export { default as Twitter } from './Twitter';
// export { default as Underline } from './Underline';
// export { default as Upload } from './Upload';
// export { default as ViewGrid } from './ViewGrid';
// export { default as ViewList } from './ViewList';
// export { default as Volume } from './Volume';
// export { default as Wheat } from './Wheat';
// export { default as World } from './World';
