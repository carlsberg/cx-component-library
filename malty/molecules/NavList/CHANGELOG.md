# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.2.0 (2024-05-31)

### Bug Fixes

- Remove `eslint` warnings and errors ([#322](https://github.com/CarlsbergGBS/cx-component-library/issues/322)) ([4c7bc73](https://github.com/CarlsbergGBS/cx-component-library/commit/4c7bc737de8abb060919ac5ba16220a6d5d0ace9))
- Run auto fixable lint based on new `eslint` configuration ([#501](https://github.com/CarlsbergGBS/cx-component-library/issues/501)) ([9ee94cc](https://github.com/CarlsbergGBS/cx-component-library/commit/9ee94cc2cd407f717e62d8857b5f20a74e7bbec4))
- **sidenav:** click on sub nav item should not call onNavItemClick ([#964](https://github.com/CarlsbergGBS/cx-component-library/issues/964)) ([d366678](https://github.com/CarlsbergGBS/cx-component-library/commit/d3666781fdec09edbc12cebd9944bd564d856ee4))
- **sidenav:** nav list test file linting error ([#971](https://github.com/CarlsbergGBS/cx-component-library/issues/971)) ([8fb53b0](https://github.com/CarlsbergGBS/cx-component-library/commit/8fb53b03d05288bcedbedfd229a450e3aba08fe5))
- **sidenav:** sidenav extract handlers from component logic ([#962](https://github.com/CarlsbergGBS/cx-component-library/issues/962)) ([70d5789](https://github.com/CarlsbergGBS/cx-component-library/commit/70d5789f86c1a3d9ba4bb06fc36d97cfbdac6261))
- Update components to have a function declaration and remove unneeded fragments ([#502](https://github.com/CarlsbergGBS/cx-component-library/issues/502)) ([b470095](https://github.com/CarlsbergGBS/cx-component-library/commit/b4700954f4e1ae9c7e547f4de37ac2b15a9d55ea))

### Features

- add tsconfig files and update package.json files for each component ([3bfc644](https://github.com/CarlsbergGBS/cx-component-library/commit/3bfc644e1cfc9dbb7cf7a0469e25fce055b53240))
- change organization name ([a0aa428](https://github.com/CarlsbergGBS/cx-component-library/commit/a0aa428f43138af5707a3ff4c0e36b7c056f02e6))
- **DSM-248:** Headline HTML tags ([#184](https://github.com/CarlsbergGBS/cx-component-library/issues/184)) ([305f5e0](https://github.com/CarlsbergGBS/cx-component-library/commit/305f5e0fa92d5813f4877b338ad827a8c347d452))
- **DSM-302:** navList button fix test ([#347](https://github.com/CarlsbergGBS/cx-component-library/issues/347)) ([bd62577](https://github.com/CarlsbergGBS/cx-component-library/commit/bd625771ab5adf948637e046b6af7f6919b36329))
- **DSM-362:** snapshots removed & unit test updates ([#307](https://github.com/CarlsbergGBS/cx-component-library/issues/307)) ([78423eb](https://github.com/CarlsbergGBS/cx-component-library/commit/78423eb5cfe04fe8a3a7af1f5af074ac6d35e235))
- **general:** [DSM-953] refactor icons implementation ([#892](https://github.com/CarlsbergGBS/cx-component-library/issues/892)) ([08d259f](https://github.com/CarlsbergGBS/cx-component-library/commit/08d259f476ce97e9db8ac41d5eb4773eabca0d37))
- **global:** [DSM-775] storybook v7 upgrade ([#782](https://github.com/CarlsbergGBS/cx-component-library/issues/782)) ([23eee21](https://github.com/CarlsbergGBS/cx-component-library/commit/23eee211673a2351970d70e9f4533f135840540b))
- **global:** migrate from bit to lerna ([a0110c8](https://github.com/CarlsbergGBS/cx-component-library/commit/a0110c8831370dc762c193b17cc593eed381f990))
- **global:** Remove `react/jsx-props-no-spreading` comments ([#558](https://github.com/CarlsbergGBS/cx-component-library/issues/558)) ([823444c](https://github.com/CarlsbergGBS/cx-component-library/commit/823444ce0800bc7f49b584de59d247eb8497f559))
- remove empty composition files ([f8173f4](https://github.com/CarlsbergGBS/cx-component-library/commit/f8173f4a2ecbf80bb7b6ffe848c023ae31819c2d))
- Several component updates ([#185](https://github.com/CarlsbergGBS/cx-component-library/issues/185)) ([633c9f0](https://github.com/CarlsbergGBS/cx-component-library/commit/633c9f0a4432e9e3d08ff05156453660929c19a2))
- Use new `eslint-config-carlsberg` with streamlined way to create components ([#559](https://github.com/CarlsbergGBS/cx-component-library/issues/559)) ([a459332](https://github.com/CarlsbergGBS/cx-component-library/commit/a45933215bc0b523220743f17d988d7f1ad5fce6))
