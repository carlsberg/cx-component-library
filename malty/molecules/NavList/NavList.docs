---
labels: ['react', 'typescript', 'Navlist']
description: 'A Nav list component'
---