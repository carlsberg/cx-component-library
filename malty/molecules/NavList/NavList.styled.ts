import styled from 'styled-components';

export const StyledNavList = styled.ul`
  width: 100%;
  list-style: none;
  padding: 0;
  margin: 0;
  box-sizing: border-box;
`;

export const StyledNavItem = styled.li<{
  selected: boolean;
}>`
  width: 100%;
  height: ${({ theme }) => theme.variables.navList.listItem.height.value}px;
  padding: ${({ theme }) => theme.variables.navList.listItem.paddingVertical.value}px
    ${({ theme }) => theme.variables.navList.listItem.paddingSide.value}px;
  text-transform: capitalize;
  position: relative;
  box-sizing: border-box;
  background-color: ${({ selected, theme }) =>
    `${selected ? theme.color.support.support80.value : theme.color.default.value}`};
  cursor: ${({ selected }) => `${selected ? 'default' : 'pointer'}`};
  user-select: none;
  & a {
    text-decoration: none;
    display: inline-block;
    width: 100%;
    height: 100%;
    position: absolute;
    top: 0;
    left: 0;
    & span {
      display: inline-block;
      position: absolute;
      left: ${({ theme }) => theme.variables.navList.listItem.labelLeft.value}px;
      top: ${({ theme }) => theme.variables.navList.listItem.labelTop.value}px;
    }
  }
  & svg {
    position: absolute;
    left: ${({ theme }) => theme.variables.navList.listItem.iconLeft.value}px;
  }
  & p {
    margin: 0;
  }
  &:hover {
    background-color: ${({ selected, theme }) => `${selected ? theme.color.support.support80.value : '#314550'}`};
    transition: background-color 0.2s ease-in-out;
  }
`;

export const StyledSubNavItem = styled(StyledNavItem)`
  padding-left: ${({ theme }) => theme.variables.navList.subItem.paddingLeft.value}px;
`;

export const StyledRightArrow = styled.span`
  & svg {
    left: initial;
    right: ${({ theme }) => theme.variables.navList.listItem.arrowRight.value}px;
    top: ${({ theme }) => theme.variables.navList.listItem.arrowTop.value}px;
  }
`;
