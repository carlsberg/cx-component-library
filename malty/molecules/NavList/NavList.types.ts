import { IconName } from '@carlsberggroup/malty.atoms.icon';
import React from 'react';

export interface SubItemProps {
  name: string;
  key?: string;
  href?: string;
  component?: React.ReactNode | JSX.Element;
}

export type ItemProps = {
  name: string;
  icon: IconName;
  key?: string;
  href?: string;
  component?: React.ReactNode | JSX.Element;
  subItems?: SubItemProps[];
  category?: string;
};

export interface NavItemProps {
  item: ItemProps;
  itemIndex: number;
  setActiveNavItem: (item: number) => void;
  openSubNav: (item: number) => void;
  selected?: boolean;
  className?: string;
}

export interface SubNavItemProps {
  item: SubItemProps;
  itemIndex: number;
  setActiveNavItem: (item: number) => void;
  selected: boolean;
}

type validCustomPropValues =
  | string
  | number
  | React.ReactNode
  | JSX.Element
  | React.MouseEventHandler<HTMLButtonElement>;

export type LinkComponentProps = {
  component?: React.ReactNode | JSX.Element;
  componentProps?: {
    [key: string]: validCustomPropValues;
  };
  href?: string;
  children?: React.ReactNode | JSX.Element;
};

export interface NavListProps {
  navItems: ItemProps[];
  activeNavItem: number;
  activeSubItem: number;
  subNavIsActive: boolean;
  setActiveNavItem: (item: number) => void;
  setActiveSubItem: (item: number) => void;
  toggleSubNav: (active: boolean) => void;
}
