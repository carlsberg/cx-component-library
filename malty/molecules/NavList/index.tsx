export { NavList } from './NavList';
export type { ItemProps, NavListProps } from './NavList.types';
