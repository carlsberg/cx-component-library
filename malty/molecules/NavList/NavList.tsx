import { CloneIcon, IconColor, IconSize } from '@carlsberggbs/malty.atoms.base-icon';
import { Text, TextColor, TextStyle } from '@carlsberggbs/malty.atoms.text';
import { ArrowSmallLeft } from '@carlsberggbs/malty.icons.arrow-small-left';
import { ArrowSmallRight } from '@carlsberggbs/malty.icons.arrow-small-right';
import { globalTheme as defaultTheme } from '@carlsberggbs/malty.theme.malty-theme-provider';
import React, { useContext } from 'react';
import { ThemeContext } from 'styled-components';
import { StyledNavItem, StyledNavList, StyledRightArrow, StyledSubNavItem } from './NavList.styled';
import { ItemProps, LinkComponentProps, NavItemProps, NavListProps, SubNavItemProps } from './NavList.types';

const LinkComponent = ({ component, href, children, componentProps }: LinkComponentProps) => {
  const CustomComponent = (component as keyof JSX.IntrinsicElements) || null;

  return component ? <CustomComponent {...componentProps}>{children}</CustomComponent> : <a href={href}>{children}</a>;
};

const SubNavItem = ({ item, onSubItemClick, selected }: SubNavItemProps) => {
  const { id, component, name, href, ...customProps } = item;
  const componentProps = { ...customProps };
  const theme = useContext(ThemeContext) || defaultTheme;

  return (
    <StyledSubNavItem onClick={() => onSubItemClick(id)} selected={selected} theme={theme} data-testid="subnav-item">
      <LinkComponent component={component} href={href} componentProps={componentProps}>
        <Text textStyle={TextStyle.MediumSmallDefault} color={TextColor.White}>
          {name}
        </Text>
      </LinkComponent>
    </StyledSubNavItem>
  );
};

const NavItem = ({ item, onNavItemClick, openSubNav, selected = false, className }: NavItemProps) => {
  const { id, component, name, href, icon, subItems, category, ...customProps } = item;
  const componentProps = { ...customProps };
  const theme = useContext(ThemeContext) || defaultTheme;
  const firstSubItemId = subItems && subItems[0]?.id;

  const handleOnClick = () => {
    if (subItems) {
      openSubNav(id, firstSubItemId as string);
    } else {
      onNavItemClick(id);
    }
  };

  return (
    <StyledNavItem
      data-testid="nav-item"
      onClick={handleOnClick}
      selected={selected}
      theme={theme}
      data-category={category}
      className={className}
    >
      <LinkComponent component={component} href={href} componentProps={componentProps}>
        <CloneIcon icon={icon} color={IconColor.White} size={IconSize.Small} />

        <Text textStyle={TextStyle.MediumSmallDefault} color={TextColor.White}>
          {name}
        </Text>

        {subItems && (
          <StyledRightArrow theme={theme}>
            <ArrowSmallRight size={IconSize.Small} color={IconColor.White} />
          </StyledRightArrow>
        )}
      </LinkComponent>
    </StyledNavItem>
  );
};

export const NavList = ({
  navItems,
  activeNavItem,
  activeSubItem,
  subNavIsActive,
  onNavItemClick,
  onSubItemClick,
  toggleSubNav
}: NavListProps) => {
  const theme = useContext(ThemeContext) || defaultTheme;

  const handleOpenSubNav = (itemId: string, firstSubItemId: string) => {
    onNavItemClick(itemId);
    toggleSubNav(!subNavIsActive);
  };

  const handleCloseSubNav = () => {
    toggleSubNav(false);
  };

  type EmptyObj = Record<PropertyKey, never>;

  const getActiveItem = (items: ItemProps[], activeItem: string | null): ItemProps | EmptyObj => {
    if (!activeNavItem) {
      return {} as ItemProps;
    }
    return items.find((item) => item.id === activeNavItem) || {};
  };

  const activeItem = getActiveItem(navItems, activeNavItem) || ({} as ItemProps);
  const { component, name, href, subItems, icon, category, ...customProps } = activeItem;
  const componentProps = { ...customProps };

  const resolveItemClass = (itemsInCategory: ItemProps[], item: ItemProps) => {
    let className;
    const isFirstInCategory = itemsInCategory.indexOf(item) === 0;
    const isLastInCategory = itemsInCategory.indexOf(item) === itemsInCategory.length - 1;

    if (isFirstInCategory) {
      className = 'firstInCategory';
    }
    if (isLastInCategory) {
      className = 'lastInCategory';
    }
    return className;
  };

  return (
    <StyledNavList theme={theme} data-testid="nav-list">
      {!subNavIsActive &&
        navItems?.map((item) => {
          const selected = activeNavItem === item.id;
          let className;

          if (item.category) {
            const itemsInCategory = navItems.filter((navItem) => navItem.category === item.category);
            // add a class to the first and the last nav items that belong to the current category
            className = resolveItemClass(itemsInCategory, item);
          }

          return (
            <NavItem
              item={item}
              onNavItemClick={onNavItemClick}
              openSubNav={handleOpenSubNav}
              selected={selected}
              key={item.key || `navItem${item.id}`}
              className={className}
            />
          );
        })}

      {subNavIsActive && (
        <>
          <StyledNavItem selected={false} onClick={handleCloseSubNav} theme={theme}>
            <LinkComponent component={component} href={href} componentProps={componentProps}>
              <ArrowSmallLeft size={IconSize.Small} color={IconColor.White} />
              <Text textStyle={TextStyle.MediumSmallBold} color={TextColor.White}>
                {name}
              </Text>
            </LinkComponent>
          </StyledNavItem>
          {subItems?.map((item, index) => {
            const selected = activeSubItem === item.id;
            return (
              <SubNavItem
                item={item}
                itemIndex={index}
                onSubItemClick={onSubItemClick}
                selected={selected}
                key={item.key || `subNavItem${index}`}
              />
            );
          })}
        </>
      )}
    </StyledNavList>
  );
};
