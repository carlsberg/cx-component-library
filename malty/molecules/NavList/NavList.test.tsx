import { DataTransfer } from '@carlsberggroup/malty.icons.data-transfer';
import { render } from '@carlsberggroup/malty.utils.test';
import { screen, within } from '@testing-library/react';
import React from 'react';
import { NavList } from './NavList';

const simpleNavigation = [
  { icon: <DataTransfer />, name: 'item 1', href: '/item1' },
  { icon: <DataTransfer />, name: 'item 2', href: '/item2' },
  { icon: <DataTransfer />, name: 'item 3', href: '/item3' }
];

const testHandler = jest.fn();

describe('navList molecule', () => {
  it('renders correct number of list items', () => {
    render(
      <NavList
        navItems={simpleNavigation}
        activeNavItem={-1}
        activeSubItem={-1}
        subNavIsActive={false}
        setActiveNavItem={testHandler}
        setActiveSubItem={testHandler}
        toggleSubNav={testHandler}
      />
    );
    const list = screen.getByRole('list');
    const { getAllByRole } = within(list);
    const items = getAllByRole('listitem');
    expect(items).toHaveLength(3);
  });
});
