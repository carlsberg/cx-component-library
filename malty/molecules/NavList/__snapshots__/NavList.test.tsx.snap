// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sideNav molecule matches snapshot 1`] = `
<ul
  className="sc-hKMtZM hWSaly"
>
  <li
    className="sc-eCYdqJ ebehEg"
    onClick={[Function]}
    selected={false}
  >
    <svg
      className="sc-gsnTZi gIwTyE"
      color="#212833"
      data-testid="svg-component"
      size="24px"
      version="1.1"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <path
        d="M16 2a1 1 0 01.117 1.993L16 4H7a1 1 0 00-.993.883L6 5v14a1 1 0 00.883.993L7 20h9a1 1 0 01.117 1.993L16 22H7a3 3 0 01-2.995-2.824L4 19V5a3 3 0 012.824-2.995L7 2zm-4 13a1 1 0 01.117 1.993L12 17H9a1 1 0 01-.117-1.993L9 15zm5.293-6.667a1 1 0 011.414 0l3 3a1 1 0 010 1.414l-3 3a1 1 0 01-1.414-1.414l2.293-2.293-2.293-2.293a1 1 0 010-1.414zM12 11a1 1 0 01.117 1.993L12 13H9a1 1 0 01-.117-1.993L9 11zm0-4a1 1 0 01.117 1.993L12 9H9a1 1 0 01-.117-1.993L9 7z"
        fillRule="evenodd"
      />
    </svg>
    <a
      href="/item1"
    />
    <span
      className="sc-bczRLJ hROJNc"
    >
      <p
        className="sc-dkzDqf jdtZuS"
        color="white"
      >
        item 1
      </p>
    </span>
  </li>
  <li
    className="sc-eCYdqJ ebehEg"
    onClick={[Function]}
    selected={false}
  >
    <svg
      className="sc-gsnTZi gIwTyE"
      color="#212833"
      data-testid="svg-component"
      size="24px"
      version="1.1"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <path
        d="M16 2a1 1 0 01.117 1.993L16 4H7a1 1 0 00-.993.883L6 5v14a1 1 0 00.883.993L7 20h9a1 1 0 01.117 1.993L16 22H7a3 3 0 01-2.995-2.824L4 19V5a3 3 0 012.824-2.995L7 2zm-4 13a1 1 0 01.117 1.993L12 17H9a1 1 0 01-.117-1.993L9 15zm5.293-6.667a1 1 0 011.414 0l3 3a1 1 0 010 1.414l-3 3a1 1 0 01-1.414-1.414l2.293-2.293-2.293-2.293a1 1 0 010-1.414zM12 11a1 1 0 01.117 1.993L12 13H9a1 1 0 01-.117-1.993L9 11zm0-4a1 1 0 01.117 1.993L12 9H9a1 1 0 01-.117-1.993L9 7z"
        fillRule="evenodd"
      />
    </svg>
    <a
      href="/item2"
    />
    <span
      className="sc-bczRLJ hROJNc"
    >
      <p
        className="sc-dkzDqf jdtZuS"
        color="white"
      >
        item 2
      </p>
    </span>
  </li>
  <li
    className="sc-eCYdqJ ebehEg"
    onClick={[Function]}
    selected={false}
  >
    <svg
      className="sc-gsnTZi gIwTyE"
      color="#212833"
      data-testid="svg-component"
      size="24px"
      version="1.1"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <path
        d="M16 2a1 1 0 01.117 1.993L16 4H7a1 1 0 00-.993.883L6 5v14a1 1 0 00.883.993L7 20h9a1 1 0 01.117 1.993L16 22H7a3 3 0 01-2.995-2.824L4 19V5a3 3 0 012.824-2.995L7 2zm-4 13a1 1 0 01.117 1.993L12 17H9a1 1 0 01-.117-1.993L9 15zm5.293-6.667a1 1 0 011.414 0l3 3a1 1 0 010 1.414l-3 3a1 1 0 01-1.414-1.414l2.293-2.293-2.293-2.293a1 1 0 010-1.414zM12 11a1 1 0 01.117 1.993L12 13H9a1 1 0 01-.117-1.993L9 11zm0-4a1 1 0 01.117 1.993L12 9H9a1 1 0 01-.117-1.993L9 7z"
        fillRule="evenodd"
      />
    </svg>
    <a
      href="/item3"
    />
    <span
      className="sc-bczRLJ hROJNc"
    >
      <p
        className="sc-dkzDqf jdtZuS"
        color="white"
      >
        item 3
      </p>
    </span>
  </li>
</ul>
`;
